homeassistant/components/livisi/coordinator.py

```diff
@@ -0,0 +1,138 @@
+"""Code to manage fetching LIVISI data API."""
+from __future__ import annotations
+
+from datetime import timedelta
+from typing import Any
+
+from aiohttp import ClientConnectorError
+from aiolivisi import AioLivisi, LivisiEvent, Websocket
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.dispatcher import async_dispatcher_send
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import (
+    AVATAR_PORT,
+    CLASSIC_PORT,
+    CONF_HOST,
+    CONF_PASSWORD,
+    DEVICE_POLLING_DELAY,
+    ID,
+    IS_REACHABLE,
+    LIVISI_REACHABILITY_CHANGE,
+    LIVISI_STATE_CHANGE,
+    LOGGER,
+    STATE,
+)
+
+
+class LivisiDataUpdateCoordinator(DataUpdateCoordinator[list[dict[str, Any]]]):
+    """Class to manage fetching LIVISI data API."""
+
+    config_entry: ConfigEntry
+
+    def __init__(
+        self, hass: HomeAssistant, config_entry: ConfigEntry, aiolivisi: AioLivisi
+    ) -> None:
+        """Initialize my coordinator."""
+        super().__init__(
+            hass,
+            LOGGER,
+            name="Livisi devices",
+            update_interval=timedelta(seconds=DEVICE_POLLING_DELAY),
+        )
+        self.config_entry = config_entry
+        self.hass = hass
+        self.aiolivisi = aiolivisi
+        self.websocket = Websocket(aiolivisi)
+        self.devices: set[str] = set()
+        self.rooms: dict[str, Any] = {}
+        self.serial_number: str = ""
+        self.controller_type: str = ""
+        self.is_avatar: bool = False
+        self.port: int = 0
+
+    async def _async_update_data(self) -> list[dict[str, Any]]:
+        """Get device configuration from LIVISI."""
+        try:
+            return await self.async_get_devices()
+        except ClientConnectorError as exc:
+            raise UpdateFailed("Failed to get LIVISI the devices") from exc
+
+    async def async_setup(self) -> None:
+        """Set up the Livisi Smart Home Controller."""
+        if not self.aiolivisi.livisi_connection_data:
+            livisi_connection_data = {
+                "ip_address": self.config_entry.data[CONF_HOST],
+                "password": self.config_entry.data[CONF_PASSWORD],
+            }
+
+            await self.aiolivisi.async_set_token(
+                livisi_connection_data=livisi_connection_data
+            )
+        controller_data = await self.aiolivisi.async_get_controller()
```

Comment:
Should we catch errors when setting the token and getting the controller?

If we want to retry later due to connection error we want to raise `ConfigEntryNotReady`. If we want to re-authenticate due to incorrect credentials, we want to raise `ConfigEntryAuthFailed` but the that requires a reauth flow in the config flow.

Adding a reauth flow can be done in a follow up PR.

https://developers.home-assistant.io/docs/config_entries_config_flow_handler#reauthentication

Reply:
I will implement this in the next pull request. Thank you. Speaking of the next pull request, we plan to add climate devices to HomeAssistant as soon as possible because winter is coming.

Reply:
I'd at least catch possible errors in this PR and raise `ConfigEntryNotReady` if we want to retry.

Reply:
I added this, thank you.

Reply:
I think you added it to the wrong place. Don't change the config flow. Update the api calls in the coordinator to handle exceptions.

During config entry setup in `async_setup_entry` in the init module we can raise `ConfigEntryNotReady` to retry the config entry setup again later.

Reply:
I understand now. I updated the code, please check if it is ok. Thank you.